rust/chains/tw_solana/src/modules/offchain_message_signer.rs

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// Copyright © 2017 Trust Wallet.
+
+use crate::SOLANA_ALPHABET;
+use tw_coin_entry::coin_context::CoinContext;
+use tw_coin_entry::error::prelude::*;
+use tw_coin_entry::modules::message_signer::MessageSigner;
+use tw_coin_entry::signing_output_error;
+use tw_encoding::base58;
+use tw_keypair::ed25519;
+use tw_keypair::traits::{SigningKeyTrait, VerifyingKeyTrait};
+use tw_misc::try_or_false;
+use tw_proto::Solana::Proto;
+use tw_proto::TxCompiler::Proto as CompilerProto;
+
+/// Currently, supports https://solana.com/developers/cookbook/wallets/sign-message only.
+pub struct OffchainMessageSigner;
+
+impl OffchainMessageSigner {
+    pub fn sign_message_impl(
+        _coin: &dyn CoinContext,
+        input: Proto::MessageSigningInput,
+    ) -> SigningResult<Proto::MessageSigningOutput<'static>> {
+        let private_key = ed25519::sha512::PrivateKey::try_from(input.private_key.as_ref())?;
+        let sign = private_key.sign(input.message.as_bytes().to_vec())?;
+        let base58_sign = base58::encode(sign.to_bytes().as_slice(), SOLANA_ALPHABET);
+        Ok(Proto::MessageSigningOutput {
+            signature: base58_sign.into(),
+            ..Proto::MessageSigningOutput::default()
+        })
+    }
+}
+
+impl MessageSigner for OffchainMessageSigner {
+    type MessageSigningInput<'a> = Proto::MessageSigningInput<'a>;
+    type MessagePreSigningOutput = CompilerProto::PreSigningOutput<'static>;
+    type MessageSigningOutput = Proto::MessageSigningOutput<'static>;
+    type MessageVerifyingInput<'a> = Proto::MessageVerifyingInput<'a>;
+
+    fn message_preimage_hashes(
+        &self,
+        _coin: &dyn CoinContext,
+        input: Self::MessageSigningInput<'_>,
+    ) -> Self::MessagePreSigningOutput {
+        CompilerProto::PreSigningOutput {
+            data_hash: input.message.as_bytes().to_vec().into(),
```

Comment:
It seems that the value is not a hash of the data. Should we assign it to the `data_hash` field?

Reply:
Fixed, thanks